# SECTION 1. IMPORTING ALL DATA INTO NOTEBOOK

In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
import warnings
warnings.filterwarnings('ignore')

**planes**

In [4]:
planes=pd.read_csv("planes.csv")

In [5]:
planes.head()

tailnum    year                     type      manufacturer      model  \
0  N10156  2004.0  Fixed wing multi engine           EMBRAER  EMB-145XR   
1  N102UW  1998.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
2  N103US  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
3  N104UW  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
4  N10575  2002.0  Fixed wing multi engine           EMBRAER  EMB-145LR   

   engines  seats  speed     engine  
0        2     55    NaN  Turbo-fan  
1        2    182    NaN  Turbo-fan  
2        2    182    NaN  Turbo-fan  
3        2    182    NaN  Turbo-fan  
4        2     55    NaN  Turbo-fan

In [6]:
planes.describe()

year      engines        seats       speed
count  3252.000000  3322.000000  3322.000000   23.000000
mean   2000.484010     1.995184   154.316376  236.782609
std       7.193425     0.117593    73.654974  149.759794
min    1956.000000     1.000000     2.000000   90.000000
25%    1997.000000     2.000000   140.000000  107.500000
50%    2001.000000     2.000000   149.000000  162.000000
75%    2005.000000     2.000000   182.000000  432.000000
max    2013.000000     4.000000   450.000000  432.000000

In [7]:
planes.isnull().sum()

tailnum            0
year              70
type               0
manufacturer       0
model              0
engines            0
seats              0
speed           3299
engine             0
dtype: int64

In [8]:
numeric_features = [feature for feature in planes.columns if planes[feature].dtype != 'O']
categorical_features = [feature for feature in planes.columns if planes[feature].dtype == 'O']
print('We have {} numerical features : {}'.format(len(numeric_features), numeric_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

We have 4 numerical features : ['year', 'engines', 'seats', 'speed']

We have 5 categorical features : ['tailnum', 'type', 'manufacturer', 'model', 'engine']


**flights**

In [9]:
flights=pd.read_csv("flights.csv",header=None)

Here first 54 rows of the data is unwanted and we drop those rows using df.drop()

In [10]:
flights=flights.drop([i for i in range(54)])

after dropping 54 rows we have unwanted data as header and we have our features(header) in first row so we make first row as our header and drop the first row

In [11]:
flights.columns = flights.iloc[0]
flights = flights.drop(flights.index[0])

In [12]:
flights.head()

54  year month day dep_time sched_dep_time dep_delay arr_time sched_arr_time  \
55  2013     1   1      517            515         2      830            819   
56  2013     1   1      533            529         4      850            830   
57  2013     1   1      542            540         2      923            850   
58  2013     1   1      544            545        -1     1004           1022   
59  2013     1   1      554            600        -6      812            837   

54 arr_delay carrier flight tailnum origin dest air_time distance hour minute  \
55        11      UA   1545  N14228    EWR  IAH      227     1400    5     15   
56        20      UA   1714  N24211    LGA  IAH      227     1416    5     29   
57        33      AA   1141  N619AA    JFK  MIA      160     1089    5     40   
58       -18      B6    725  N804JB    JFK  BQN      183     1576    5     45   
59       -25      DL    461  N668DN    LGA  ATL      116      762    6      0   

54         time_hour  
55  01-01-2013 05:00  
56  01-01-2013 05:00  
57  01-01-2013 05:00  
58  01-01-2013 05:00  
59  01-01-2013 06:00

In [13]:
flights.isnull().sum()

54
year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
dtype: int64

In [14]:
len(flights)

336776

**airlines**

In [15]:
airlines=pd.read_csv("airlines.csv")

In [16]:
airlines

carrier                         name
0       9E            Endeavor Air Inc.
1       AA       American Airlines Inc.
2       AS         Alaska Airlines Inc.
3       B6              JetBlue Airways
4       DL         Delta Air Lines Inc.
5       EV     ExpressJet Airlines Inc.
6       F9       Frontier Airlines Inc.
7       FL  AirTran Airways Corporation
8       HA       Hawaiian Airlines Inc.
9       MQ                    Envoy Air
10      OO        SkyWest Airlines Inc.
11      UA        United Air Lines Inc.
12      US              US Airways Inc.
13      VX               Virgin America
14      WN       Southwest Airlines Co.
15      YV           Mesa Airlines Inc.

In [17]:
len(airlines)

16

**weather**

In [18]:
weather=pd.read_csv('weather.csv')

In [19]:
weather.head()

origin  year  month  day  hour   temp   dewp  humid  wind_dir  wind_speed  \
0    EWR  2013      1    1     0  37.04  21.92  53.97     230.0    10.35702   
1    EWR  2013      1    1     1  37.04  21.92  53.97     230.0    13.80936   
2    EWR  2013      1    1     2  37.94  21.92  52.09     230.0    12.65858   
3    EWR  2013      1    1     3  37.94  23.00  54.51     230.0    13.80936   
4    EWR  2013      1    1     4  37.94  24.08  57.04     240.0    14.96014   

   wind_gust  precip  pressure  visib            time_hour  
0  11.918651     0.0    1013.9   10.0  2013-01-01 01:00:00  
1  15.891535     0.0    1013.0   10.0  2013-01-01 02:00:00  
2  14.567241     0.0    1012.6   10.0  2013-01-01 03:00:00  
3  15.891535     0.0    1012.7   10.0  2013-01-01 04:00:00  
4  17.215830     0.0    1012.8   10.0  2013-01-01 05:00:00

**QUERIES**

 # 1.SELECT DISTINCT engine FROM planes

In [20]:
planes['engine'].nunique()

6

SIX DISTINCT ENGINE IN PLANES

In [21]:
result_1=planes['engine'].unique()

In [22]:
result_1

array(['Turbo-fan', 'Turbo-jet', 'Reciprocating', '4 Cycle',
       'Turbo-shaft', 'Turbo-prop'], dtype=object)

In [23]:
len(result_1)

6

**HERE WE NEED TO FIND  UNIQUE ENGINE IN GIVEN DATA,GIVEN DATA CONSISTS 6 UNIQUE ENGINE AND THEY ARE  'Turbo-fan', 'Turbo-jet', 'Reciprocating', '4 Cycle','Turbo-shaft', 'Turbo-prop'**

# 2.SELECT DISTINCT type, engine FROM planes

**way 1**

In [24]:
set_2=set([])
for i in range(len(planes)):
    set_2.add((planes['type'][i],planes['engine'][i]))    

In [25]:
print("type                        engine")
for x,y in set_2:
    print(x,"   ",y)

type                        engine
Fixed wing multi engine     Reciprocating
Fixed wing single engine     4 Cycle
Fixed wing multi engine     Turbo-fan
Fixed wing multi engine     Turbo-jet
Rotorcraft     Turbo-shaft
Fixed wing single engine     Reciprocating
Fixed wing multi engine     Turbo-prop


 **way 2**

In [26]:
result_2 = planes[['type', 'engine']].drop_duplicates()
print(len(result_2))
result_2

7


type         engine
0      Fixed wing multi engine      Turbo-fan
51     Fixed wing multi engine      Turbo-jet
424   Fixed wing single engine  Reciprocating
427    Fixed wing multi engine  Reciprocating
686   Fixed wing single engine        4 Cycle
811                 Rotorcraft    Turbo-shaft
1045   Fixed wing multi engine     Turbo-prop

In [27]:
result_2.shape

(7, 2)

**Here we need to find different pairs of type,engine pairs in given data and we have 7 such distinct pairs as shown  above**

# 3 SELECT COUNT(*), engine FROM planes GROUP BY engine

In [28]:
result_3= planes.groupby('engine').size().reset_index(name='COUNT(*)')

In [29]:
result_3

engine  COUNT(*)
0        4 Cycle         2
1  Reciprocating        28
2      Turbo-fan      2750
3      Turbo-jet       535
4     Turbo-prop         2
5    Turbo-shaft         5

In [30]:
result_3.shape

(6, 2)

**Here in given Data we have 2 4 Cycle engine,28 Reciprocating engine ,2750 Turbo-fan engine,535 Turbo-jet engine ,2 Turbo-prop engine and 5 Turbo-shaft engine**

# 4. SELECT COUNT(*), engine, type FROM planes GROUP BY engine, type

In [31]:
result_4= planes.groupby(['engine', 'type']).size().reset_index(name='COUNT(*)')

In [32]:
result_4

engine                      type  COUNT(*)
0        4 Cycle  Fixed wing single engine         2
1  Reciprocating   Fixed wing multi engine         5
2  Reciprocating  Fixed wing single engine        23
3      Turbo-fan   Fixed wing multi engine      2750
4      Turbo-jet   Fixed wing multi engine       535
5     Turbo-prop   Fixed wing multi engine         2
6    Turbo-shaft                Rotorcraft         5

In [33]:
result_4.shape

(7, 3)

**Here we need to find engine,type pairs and their respective counts**

# 5.SELECT MIN(year),AVG(year), MAX(year), engine,manufacturer FROM planes GROUP BY engine ,  manufacturer

In [34]:
group_by= planes.groupby(['engine', 'manufacturer']).size().reset_index(name="COUNT(*)")
group_by

engine                   manufacturer  COUNT(*)
0         4 Cycle                         CESSNA         1
1         4 Cycle                    JOHN G HESS         1
2   Reciprocating          AMERICAN AIRCRAFT INC         2
3   Reciprocating             AVIAT AIRCRAFT INC         1
4   Reciprocating                  BARKER JACK L         1
5   Reciprocating                         CESSNA         7
6   Reciprocating             CIRRUS DESIGN CORP         1
7   Reciprocating                    DEHAVILLAND         1
8   Reciprocating                        DOUGLAS         1
9   Reciprocating                 FRIEDEMANN JON         1
10  Reciprocating             HURLEY JAMES LARRY         1
11  Reciprocating                   KILDALL GARY         1
12  Reciprocating                LAMBERT RICHARD         1
13  Reciprocating                LEBLANC GLENN T         1
14  Reciprocating                     MARZ BARRY         1
15  Reciprocating                    PAIR MIKE E         1
16  Reciprocating                          PIPER         5
17  Reciprocating                   STEWART MACO         2
18      Turbo-fan                         AIRBUS       331
19      Turbo-fan               AIRBUS INDUSTRIE       270
20      Turbo-fan         AVIONS MARCEL DASSAULT         1
21      Turbo-fan                         BOEING      1276
22      Turbo-fan                 BOMBARDIER INC       368
23      Turbo-fan                       CANADAIR         9
24      Turbo-fan                         CESSNA         1
25      Turbo-fan                        EMBRAER       298
26      Turbo-fan           GULFSTREAM AEROSPACE         1
27      Turbo-fan                    LEARJET INC         1
28      Turbo-fan              MCDONNELL DOUGLAS        91
29      Turbo-fan  MCDONNELL DOUGLAS AIRCRAFT CO       103
30      Turbo-jet                         AIRBUS         5
31      Turbo-jet               AIRBUS INDUSTRIE       130
32      Turbo-jet                         BOEING       354
33      Turbo-jet                   CANADAIR LTD         1
34      Turbo-jet                        EMBRAER         1
35      Turbo-jet           GULFSTREAM AEROSPACE         1
36      Turbo-jet              MCDONNELL DOUGLAS        29
37      Turbo-jet  MCDONNELL DOUGLAS CORPORATION        14
38     Turbo-prop                          BEECH         2
39    Turbo-shaft                     AGUSTA SPA         1
40    Turbo-shaft                           BELL         2
41    Turbo-shaft         ROBINSON HELICOPTER CO         1
42    Turbo-shaft                       SIKORSKY         1

In [35]:
result_5= planes.groupby(['engine', 'manufacturer']).agg({'year': ['min', 'mean', 'max']}).reset_index()
result_5.columns = ['engine', 'manufacturer', 'MIN(year)', 'AVG(year)', 'MAX(year)']
result_5

engine                   manufacturer  MIN(year)    AVG(year)  \
0         4 Cycle                         CESSNA     1975.0  1975.000000   
1         4 Cycle                    JOHN G HESS        NaN          NaN   
2   Reciprocating          AMERICAN AIRCRAFT INC        NaN          NaN   
3   Reciprocating             AVIAT AIRCRAFT INC     2007.0  2007.000000   
4   Reciprocating                  BARKER JACK L        NaN          NaN   
5   Reciprocating                         CESSNA     1959.0  1971.142857   
6   Reciprocating             CIRRUS DESIGN CORP     2007.0  2007.000000   
7   Reciprocating                    DEHAVILLAND     1959.0  1959.000000   
8   Reciprocating                        DOUGLAS     1956.0  1956.000000   
9   Reciprocating                 FRIEDEMANN JON     2007.0  2007.000000   
10  Reciprocating             HURLEY JAMES LARRY        NaN          NaN   
11  Reciprocating                   KILDALL GARY     1985.0  1985.000000   
12  Reciprocating                LAMBERT RICHARD        NaN          NaN   
13  Reciprocating                LEBLANC GLENN T     1985.0  1985.000000   
14  Reciprocating                     MARZ BARRY     1993.0  1993.000000   
15  Reciprocating                    PAIR MIKE E        NaN          NaN   
16  Reciprocating                          PIPER     1968.0  1976.400000   
17  Reciprocating                   STEWART MACO     1985.0  1985.000000   
18      Turbo-fan                         AIRBUS     2002.0  2007.281734   
19      Turbo-fan               AIRBUS INDUSTRIE     1992.0  1998.599251   
20      Turbo-fan         AVIONS MARCEL DASSAULT     1986.0  1986.000000   
21      Turbo-fan                         BOEING     1965.0  2001.210400   
22      Turbo-fan                 BOMBARDIER INC     1998.0  2004.486188   
23      Turbo-fan                       CANADAIR     1997.0  1997.333333   
24      Turbo-fan                         CESSNA     1979.0  1979.000000   
25      Turbo-fan                        EMBRAER     1998.0  2003.613014   
26      Turbo-fan           GULFSTREAM AEROSPACE     1992.0  1992.000000   
27      Turbo-fan                    LEARJET INC        NaN          NaN   
28      Turbo-fan              MCDONNELL DOUGLAS     1986.0  1989.704545   
29      Turbo-fan  MCDONNELL DOUGLAS AIRCRAFT CO     1987.0  1989.737864   
30      Turbo-jet                         AIRBUS     2002.0  2002.000000   
31      Turbo-jet               AIRBUS INDUSTRIE     1989.0  1997.439024   
32      Turbo-jet                         BOEING     1984.0  1996.368272   
33      Turbo-jet                   CANADAIR LTD     1974.0  1974.000000   
34      Turbo-jet                        EMBRAER     1999.0  1999.000000   
35      Turbo-jet           GULFSTREAM AEROSPACE     1976.0  1976.000000   
36      Turbo-jet              MCDONNELL DOUGLAS     1975.0  1990.714286   
37      Turbo-jet  MCDONNELL DOUGLAS CORPORATION     1991.0  1991.928571   
38     Turbo-prop                          BEECH     1967.0  1969.500000   
39    Turbo-shaft                     AGUSTA SPA     2001.0  2001.000000   
40    Turbo-shaft                           BELL     1975.0  1984.500000   
41    Turbo-shaft         ROBINSON HELICOPTER CO     2012.0  2012.000000   
42    Turbo-shaft                       SIKORSKY     1985.0  1985.000000   

    MAX(year)  
0      1975.0  
1         NaN  
2         NaN  
3      2007.0  
4         NaN  
5      1983.0  
6      2007.0  
7      1959.0  
8      1956.0  
9      2007.0  
10        NaN  
11     1985.0  
12        NaN  
13     1985.0  
14     1993.0  
15        NaN  
16     1980.0  
17     1985.0  
18     2013.0  
19     2013.0  
20     1986.0  
21     2013.0  
22     2013.0  
23     1998.0  
24     1979.0  
25     2013.0  
26     1992.0  
27        NaN  
28     1997.0  
29     1993.0  
30     2002.0  
31     2004.0  
32     2005.0  
33     1974.0  
34     1999.0  
35     1976.0  
36     1998.0  
37     1992.0  
38     1972.0  
39     2001.0  
40     199

In [36]:
result_5.shape

(43, 5)

**Here We can See that min(year),avg(year),max(year) have same values,because we have only one such type of engine,manufacture pair**

# 6.SELECT * FROM planes WHERE speed IS NOT NULLS

In [37]:
planes['speed'].notnull()

0       False
1       False
2       False
3       False
4       False
        ...  
3317    False
3318    False
3319    False
3320    False
3321    False
Name: speed, Length: 3322, dtype: bool

Here .notnull() function checks every row value of given column and return "False" if it is null and "True" if it is notnull(have some value).


**Here we asked to consider only not null values means True values, by passing this (planes['speed'].notnull()) result to data_frame(planes) it only consider True Value Rows and Neglect All the NULL VALUED Rows**

In [38]:
result_6=planes[planes['speed'].notnull()]

In [39]:
result_6

tailnum    year                      type       manufacturer  \
424   N201AA  1959.0  Fixed wing single engine             CESSNA   
427   N202AA  1980.0   Fixed wing multi engine             CESSNA   
821   N350AA  1980.0   Fixed wing multi engine              PIPER   
893   N364AA  1973.0   Fixed wing multi engine             CESSNA   
1027  N378AA  1963.0  Fixed wing single engine             CESSNA   
1037  N381AA  1956.0   Fixed wing multi engine            DOUGLAS   
1190  N425AA  1968.0  Fixed wing single engine              PIPER   
1430  N508AA  1975.0                Rotorcraft               BELL   
1480  N519MQ  1983.0  Fixed wing single engine             CESSNA   
1515  N525AA  1980.0   Fixed wing multi engine              PIPER   
1589  N545AA  1976.0  Fixed wing single engine              PIPER   
1694  N567AA  1959.0  Fixed wing single engine        DEHAVILLAND   
1813  N600TR  1979.0   Fixed wing multi engine  MCDONNELL DOUGLAS   
1867  N615AA  1967.0   Fixed wing multi engine              BEECH   
1883  N621AA  1975.0  Fixed wing single engine             CESSNA   
2131  N675MC  1975.0   Fixed wing multi engine  MCDONNELL DOUGLAS   
2309  N737MQ  1977.0  Fixed wing single engine             CESSNA   
2402  N762NC  1976.0   Fixed wing multi engine  MCDONNELL DOUGLAS   
2432  N767NC  1977.0   Fixed wing multi engine  MCDONNELL DOUGLAS   
2472  N774NC  1978.0   Fixed wing multi engine  MCDONNELL DOUGLAS   
2483  N777NC  1979.0   Fixed wing multi engine  MCDONNELL DOUGLAS   
2492  N779NC  1979.0   Fixed wing multi engine  MCDONNELL DOUGLAS   
2503  N782NC  1980.0   Fixed wing multi engine  MCDONNELL DOUGLAS   

            model  engines  seats  speed         engine  
424           150        1      2   90.0  Reciprocating  
427          421C        2      8   90.0  Reciprocating  
821     PA-31-350        2      8  162.0  Reciprocating  
893          310Q        2      6  167.0  Reciprocating  
1027         172E        1      4  105.0  Reciprocating  
1037       DC-7BF        4    102  232.0  Reciprocating  
1190    PA-28-180        1      4  107.0  Reciprocating  
1430         206B        1      5  112.0    Turbo-shaft  
1480        A185F        1      6  127.0  Reciprocating  
1515    PA-31-350        2      8  162.0  Reciprocating  
1589   PA-32R-300        1      7  126.0  Reciprocating  
1694  OTTER DHC-3        1     16   95.0  Reciprocating  
1813      DC-9-51        2    139  432.0      Turbo-jet  
1867       65-A90        2      9  202.0     Turbo-prop  
1883         172M        1      4  108.0        4 Cycle  
2131      DC-9-51        2    139  432.0      Turbo-jet  
2309         172N        1      4  105.0  Reciprocating  
2402      DC-9-51        2    139  432.0      Turbo-jet  
2432      DC-9-51        2    139  432.0      Turbo-jet  
2472      DC-9-51        2    139  432.0      Turbo-jet  
2483      DC-9-51        2    139  432.0      Turbo-jet  
2492      DC-9-51        2    139  432.0      Turbo-jet  
2503      DC-9-51        2    139  432.0      Turbo-jet

In [40]:
result_6.shape

(23, 9)

In [41]:
len(result_6),len(planes),len(planes[planes['speed'].isnull()]) # .isnull() returns null value counts

(23, 3322, 3299)

**we have overall 3322 ROWS AND SPEED HAS 3299 VALUES SO WE HAVE LEFT WITH (3322-3299) 23 rows** 

# 7. SELECT tailnum FROM planes WHERE seats BETWEEN 150 AND 190 AND year >= 2012


**First we will filter the planes having seats between 150 and and 190 and planes which present are after year 2011**

In [42]:
filtered = planes[(planes['seats'] >= 150) & (planes['seats'] <= 190) & (planes['year'] >= 2012)]

In [43]:
filtered

tailnum    year                     type manufacturer     model  engines  \
875   N361VA  2013.0  Fixed wing multi engine       AIRBUS  A320-214        2   
2745  N849VA  2012.0  Fixed wing multi engine       AIRBUS  A320-214        2   
2754  N851VA  2012.0  Fixed wing multi engine       AIRBUS  A320-214        2   
2757  N852VA  2012.0  Fixed wing multi engine       AIRBUS  A320-214        2   
2761  N853VA  2012.0  Fixed wing multi engine       AIRBUS  A320-214        2   
2766  N854VA  2012.0  Fixed wing multi engine       AIRBUS  A320-214        2   
2769  N855VA  2012.0  Fixed wing multi engine       AIRBUS  A320-214        2   

      seats  speed     engine  
875     182    NaN  Turbo-fan  
2745    182    NaN  Turbo-fan  
2754    182    NaN  Turbo-fan  
2757    182    NaN  Turbo-fan  
2761    182    NaN  Turbo-fan  
2766    182    NaN  Turbo-fan  
2769    182    NaN  Turbo-fan

from this filtered columns we only need 'tailnum' of planes

In [44]:
result_7=filtered['tailnum']

In [45]:
result_7.to_frame()

tailnum
875   N361VA
2745  N849VA
2754  N851VA
2757  N852VA
2761  N853VA
2766  N854VA
2769  N855VA

In [46]:
result_7.shape

(7,)

**Here we are asked to select tailnum which are having seats count between 150 and 190 and after the year 2011 i.e,from 2012 onwards**

# 8.SELECT tailnum,manufacturer,seats FROM planes WHERE manufacturer IN("BOEING", "AIRBUS", "EMBRAER") AND seats>390

**first we need to select manufacture compaies whic are in BOEING,AIRBUS,EMBRAER and seats count greater than 390**

In [47]:
filtered =planes[(planes['manufacturer'].isin(['BOEING', 'AIRBUS', 'EMBRAER'])) & (planes['seats'] > 390)]

In [48]:
filtered

tailnum    year                     type manufacturer    model  engines  \
439   N206UA  1999.0  Fixed wing multi engine       BOEING  777-222        2   
484   N228UA  2002.0  Fixed wing multi engine       BOEING  777-222        2   
577   N272AT     NaN  Fixed wing multi engine       BOEING  777-200        2   
1708  N57016  2000.0  Fixed wing multi engine       BOEING  777-224        2   
2109  N670US  1990.0  Fixed wing multi engine       BOEING  747-451        4   
2441  N77012  1999.0  Fixed wing multi engine       BOEING  777-224        2   
2485  N777UA  1995.0  Fixed wing multi engine       BOEING  777-222        2   
2494  N78003  1998.0  Fixed wing multi engine       BOEING  777-224        2   
2495  N78013  1999.0  Fixed wing multi engine       BOEING  777-224        2   
2519  N787UA  1997.0  Fixed wing multi engine       BOEING  777-222        2   
2804  N862DA  1999.0  Fixed wing multi engine       BOEING  777-232        2   
2806  N863DA  1999.0  Fixed wing multi engine       BOEING  777-232        2   
2809  N865DA  1999.0  Fixed wing multi engine       BOEING  777-232        2   

      seats  speed     engine  
439     400    NaN  Turbo-fan  
484     400    NaN  Turbo-fan  
577     400    NaN  Turbo-jet  
1708    400    NaN  Turbo-fan  
2109    450    NaN  Turbo-jet  
2441    400    NaN  Turbo-fan  
2485    400    NaN  Turbo-fan  
2494    400    NaN  Turbo-fan  
2495    400    NaN  Turbo-fan  
2519    400    NaN  Turbo-fan  
2804    400    NaN  Turbo-jet  
2806    400    NaN  Turbo-jet  
2809    400    NaN  Turbo-jet

**from the filtered columns we will select only required columns**

In [49]:
result_8=filtered[['tailnum', 'manufacturer', 'seats']]

In [50]:
result_8

tailnum manufacturer  seats
439   N206UA       BOEING    400
484   N228UA       BOEING    400
577   N272AT       BOEING    400
1708  N57016       BOEING    400
2109  N670US       BOEING    450
2441  N77012       BOEING    400
2485  N777UA       BOEING    400
2494  N78003       BOEING    400
2495  N78013       BOEING    400
2519  N787UA       BOEING    400
2804  N862DA       BOEING    400
2806  N863DA       BOEING    400
2809  N865DA       BOEING    400

In [51]:
result_8.shape

(13, 3)

# 9.SELECT DISTINCT year,seats FROM planes WHERE year >= 2012 ORDER BY year ASC,seats DESC

**first we well find the number of planes after the year 2011 or (year>=2012)**

In [52]:
planes_year_greater_than_2012=planes[planes['year'] >= 2012]

In [53]:
planes_year_greater_than_2012

tailnum    year                     type      manufacturer     model  \
215   N150UW  2013.0  Fixed wing multi engine            AIRBUS  A321-211   
216   N151UW  2013.0  Fixed wing multi engine            AIRBUS  A321-211   
218   N152UW  2013.0  Fixed wing multi engine            AIRBUS  A321-211   
221   N153UW  2013.0  Fixed wing multi engine            AIRBUS  A321-211   
223   N154UW  2013.0  Fixed wing multi engine            AIRBUS  A321-211   
...      ...     ...                      ...               ...       ...   
2801  N8620H  2013.0  Fixed wing multi engine            BOEING   737-8H4   
2802  N8621A  2013.0  Fixed wing multi engine            BOEING   737-8H4   
2919  N903JB  2013.0  Fixed wing multi engine            AIRBUS  A321-231   
2945  N907JB  2013.0  Fixed wing multi engine  AIRBUS INDUSTRIE  A321-231   
2982  N913JB  2013.0  Fixed wing multi engine            AIRBUS  A321-231   

      engines  seats  speed     engine  
215         2    199    NaN  Turbo-fan  
216         2    199    NaN  Turbo-fan  
218         2    199    NaN  Turbo-fan  
221         2    199    NaN  Turbo-fan  
223         2    199    NaN  Turbo-fan  
...       ...    ...    ...        ...  
2801        2    140    NaN  Turbo-fan  
2802        2    140    NaN  Turbo-fan  
2919        2    379    NaN  Turbo-fan  
2945        2    379    NaN  Turbo-fan  
2982        2    379    NaN  Turbo-fan  

[187 rows x 9 columns]

**from those filtered planes we need to consider two rows row and year**

In [54]:
year_and_seats_pairs=planes_year_greater_than_2012[['year', 'seats']]

In [55]:
year_and_seats_pairs

year  seats
215   2013.0    199
216   2013.0    199
218   2013.0    199
221   2013.0    199
223   2013.0    199
...      ...    ...
2801  2013.0    140
2802  2013.0    140
2919  2013.0    379
2945  2013.0    379
2982  2013.0    379

[187 rows x 2 columns]

**from this pairs we need to consider just distinct so we need to drop duplicates**

In [56]:
unique_pairs=year_and_seats_pairs.drop_duplicates()

In [57]:
unique_pairs

year  seats
215   2013.0    199
445   2012.0    260
578   2013.0     95
584   2013.0    191
609   2012.0    191
815   2012.0     20
843   2013.0     20
875   2013.0    182
1065  2012.0    377
1083  2013.0    377
1128  2012.0    222
1147  2013.0    222
1559  2012.0    149
1564  2012.0      5
1637  2012.0    379
1696  2013.0    379
2539  2012.0    200
2618  2013.0    200
2630  2012.0    140
2745  2012.0    182
2787  2013.0    140

In [58]:
unique_pairs.shape

(21, 2)

**we need to sort this unique pairs of years and seats by ascending order of year and descending order of seats**

In [59]:
result_9=unique_pairs.sort_values(by=['year', 'seats'], ascending=[True, False])

In [60]:
result_9

year  seats
1637  2012.0    379
1065  2012.0    377
445   2012.0    260
1128  2012.0    222
2539  2012.0    200
609   2012.0    191
2745  2012.0    182
1559  2012.0    149
2630  2012.0    140
815   2012.0     20
1564  2012.0      5
1696  2013.0    379
1083  2013.0    377
1147  2013.0    222
2618  2013.0    200
215   2013.0    199
584   2013.0    191
875   2013.0    182
2787  2013.0    140
578   2013.0     95
843   2013.0     20

In [61]:
result_9.shape

(21, 2)

# 10.SELECT DISTINCT year,seats FROM planes WHERE year >= 2012 ORDER BY seats DESC, year ASC

**Here first ordere of seats is considered followed by order of year**

In [62]:
result_10=unique_pairs.sort_values(by=['seats', 'year'], ascending=[False, True])

In [63]:
result_10

year  seats
1637  2012.0    379
1696  2013.0    379
1065  2012.0    377
1083  2013.0    377
445   2012.0    260
1128  2012.0    222
1147  2013.0    222
2539  2012.0    200
2618  2013.0    200
215   2013.0    199
609   2012.0    191
584   2013.0    191
2745  2012.0    182
875   2013.0    182
1559  2012.0    149
2630  2012.0    140
2787  2013.0    140
578   2013.0     95
815   2012.0     20
843   2013.0     20
1564  2012.0      5

**the above querey is similar to 9Th query But the only difference is Here first we need to consider the order of seats and followed by the order of year**

In [64]:
result_10.shape

(21, 2)

# 11. SELECT manufacturer, COUNT(*) FROM planes WHERE seats > 200 GROUP BY manufacturer

**first we will consider planes the number of seats greater than 200**

In [65]:
planes_seats_g_200=planes[planes['seats']>200]

In [66]:
planes_seats_g_200

tailnum    year                     type      manufacturer     model  \
49    N1200K  1998.0  Fixed wing multi engine            BOEING   767-332   
50    N1201P  1998.0  Fixed wing multi engine            BOEING   767-332   
69    N121DE  1987.0  Fixed wing multi engine            BOEING   767-332   
117   N136DL  1991.0  Fixed wing multi engine            BOEING   767-332   
122   N137DL  1991.0  Fixed wing multi engine            BOEING   767-332   
...      ...     ...                      ...               ...       ...   
2806  N863DA  1999.0  Fixed wing multi engine            BOEING   777-232   
2809  N865DA  1999.0  Fixed wing multi engine            BOEING   777-232   
2919  N903JB  2013.0  Fixed wing multi engine            AIRBUS  A321-231   
2945  N907JB  2013.0  Fixed wing multi engine  AIRBUS INDUSTRIE  A321-231   
2982  N913JB  2013.0  Fixed wing multi engine            AIRBUS  A321-231   

      engines  seats  speed     engine  
49          2    330    NaN  Turbo-fan  
50          2    330    NaN  Turbo-fan  
69          2    330    NaN  Turbo-fan  
117         2    330    NaN  Turbo-fan  
122         2    330    NaN  Turbo-fan  
...       ...    ...    ...        ...  
2806        2    400    NaN  Turbo-jet  
2809        2    400    NaN  Turbo-jet  
2919        2    379    NaN  Turbo-fan  
2945        2    379    NaN  Turbo-fan  
2982        2    379    NaN  Turbo-fan  

[295 rows x 9 columns]

**we need to  gruop the above by manufacturer and show their respective count**

In [67]:
result_11=planes_seats_g_200.groupby(['manufacturer']).size().reset_index(name="COUNT(*)")

In [68]:
result_11

manufacturer  COUNT(*)
0            AIRBUS        66
1  AIRBUS INDUSTRIE         4
2            BOEING       225

In [69]:
result_11.shape

(3, 2)

# 12.SELECT manufacturer, COUNT(*) FROM planes GROUP BY manufacturer HAVING COUNT(*) > 10

In [70]:
manufacturer_group=planes.groupby(['manufacturer']).size().reset_index(name='COUNT()')

In [71]:
manufacturer_group

manufacturer  COUNT()
0                      AGUSTA SPA        1
1                          AIRBUS      336
2                AIRBUS INDUSTRIE      400
3           AMERICAN AIRCRAFT INC        2
4              AVIAT AIRCRAFT INC        1
5          AVIONS MARCEL DASSAULT        1
6                   BARKER JACK L        1
7                           BEECH        2
8                            BELL        2
9                          BOEING     1630
10                 BOMBARDIER INC      368
11                       CANADAIR        9
12                   CANADAIR LTD        1
13                         CESSNA        9
14             CIRRUS DESIGN CORP        1
15                    DEHAVILLAND        1
16                        DOUGLAS        1
17                        EMBRAER      299
18                 FRIEDEMANN JON        1
19           GULFSTREAM AEROSPACE        2
20             HURLEY JAMES LARRY        1
21                    JOHN G HESS        1
22                   KILDALL GARY        1
23                LAMBERT RICHARD        1
24                    LEARJET INC        1
25                LEBLANC GLENN T        1
26                     MARZ BARRY        1
27              MCDONNELL DOUGLAS      120
28  MCDONNELL DOUGLAS AIRCRAFT CO      103
29  MCDONNELL DOUGLAS CORPORATION       14
30                    PAIR MIKE E        1
31                          PIPER        5
32         ROBINSON HELICOPTER CO        1
33                       SIKORSKY        1
34                   STEWART MACO        2

**from the above manufacturer groups consider groups having count greater than 10**

In [72]:
result_12=manufacturer_group[manufacturer_group['COUNT()']>10]

In [73]:
result_12

manufacturer  COUNT()
1                          AIRBUS      336
2                AIRBUS INDUSTRIE      400
9                          BOEING     1630
10                 BOMBARDIER INC      368
17                        EMBRAER      299
27              MCDONNELL DOUGLAS      120
28  MCDONNELL DOUGLAS AIRCRAFT CO      103
29  MCDONNELL DOUGLAS CORPORATION       14

In [74]:
result_12.shape

(8, 2)

# 13.SELECTmanufacturer, COUNT(*) FROM planes WHERE seats > 200 GROUP BY manufacturer HAVING COUNT(*) > 10

**this problem is same as 11th query and the change is we need to consider manufacturer having count greater than 10**

In [75]:
result_13=result_11[result_11['COUNT(*)']>10]

In [76]:
result_13

manufacturer  COUNT(*)
0       AIRBUS        66
2       BOEING       225

In [77]:
result_13.shape

(2, 2)

# 14. SELECT manufacturer, COUNT(*) AS howmany FROM planes GROUP BY manufacturer ORDER BY howmany DESC LIMIT 5

**In this query we have to consider manufacturer groups and store the count as individual manufacturer as howmany and order these groups by considering descending order of howmany(count(*)) and from this we need to consider top 5**

In [78]:
manufacturer_group=planes.groupby(['manufacturer']).size().reset_index(name='howmany')

In [79]:
manufacturer_group

manufacturer  howmany
0                      AGUSTA SPA        1
1                          AIRBUS      336
2                AIRBUS INDUSTRIE      400
3           AMERICAN AIRCRAFT INC        2
4              AVIAT AIRCRAFT INC        1
5          AVIONS MARCEL DASSAULT        1
6                   BARKER JACK L        1
7                           BEECH        2
8                            BELL        2
9                          BOEING     1630
10                 BOMBARDIER INC      368
11                       CANADAIR        9
12                   CANADAIR LTD        1
13                         CESSNA        9
14             CIRRUS DESIGN CORP        1
15                    DEHAVILLAND        1
16                        DOUGLAS        1
17                        EMBRAER      299
18                 FRIEDEMANN JON        1
19           GULFSTREAM AEROSPACE        2
20             HURLEY JAMES LARRY        1
21                    JOHN G HESS        1
22                   KILDALL GARY        1
23                LAMBERT RICHARD        1
24                    LEARJET INC        1
25                LEBLANC GLENN T        1
26                     MARZ BARRY        1
27              MCDONNELL DOUGLAS      120
28  MCDONNELL DOUGLAS AIRCRAFT CO      103
29  MCDONNELL DOUGLAS CORPORATION       14
30                    PAIR MIKE E        1
31                          PIPER        5
32         ROBINSON HELICOPTER CO        1
33                       SIKORSKY        1
34                   STEWART MACO        2

**now sort the above according to howmany**

In [80]:
sorted_group=manufacturer_group.sort_values(by=['howmany'], ascending=[False])

In [81]:
sorted_group

manufacturer  howmany
9                          BOEING     1630
2                AIRBUS INDUSTRIE      400
10                 BOMBARDIER INC      368
1                          AIRBUS      336
17                        EMBRAER      299
27              MCDONNELL DOUGLAS      120
28  MCDONNELL DOUGLAS AIRCRAFT CO      103
29  MCDONNELL DOUGLAS CORPORATION       14
13                         CESSNA        9
11                       CANADAIR        9
31                          PIPER        5
19           GULFSTREAM AEROSPACE        2
34                   STEWART MACO        2
8                            BELL        2
3           AMERICAN AIRCRAFT INC        2
7                           BEECH        2
12                   CANADAIR LTD        1
25                LEBLANC GLENN T        1
33                       SIKORSKY        1
32         ROBINSON HELICOPTER CO        1
30                    PAIR MIKE E        1
4              AVIAT AIRCRAFT INC        1
5          AVIONS MARCEL DASSAULT        1
6                   BARKER JACK L        1
26                     MARZ BARRY        1
24                    LEARJET INC        1
14             CIRRUS DESIGN CORP        1
23                LAMBERT RICHARD        1
22                   KILDALL GARY        1
21                    JOHN G HESS        1
20             HURLEY JAMES LARRY        1
18                 FRIEDEMANN JON        1
16                        DOUGLAS        1
15                    DEHAVILLAND        1
0                      AGUSTA SPA        1

**we need to consider top 5 groups**

In [82]:
result_14=sorted_group.head(5)

In [83]:
result_14

manufacturer  howmany
9             BOEING     1630
2   AIRBUS INDUSTRIE      400
10    BOMBARDIER INC      368
1             AIRBUS      336
17           EMBRAER      299

In [84]:
result_14.shape

(5, 2)

# **15.SELECT flights.*, planes.year AS plane_year, planes.speed AS plane_speed, planes.seats  AS plane_seats FROM flights LEFT JOIN planes ON flights.tailnum=planes.tailnum**

**GIVEN PRIMARY KEY AS TAILNUM**

In [85]:
flights.head()

54  year month day dep_time sched_dep_time dep_delay arr_time sched_arr_time  \
55  2013     1   1      517            515         2      830            819   
56  2013     1   1      533            529         4      850            830   
57  2013     1   1      542            540         2      923            850   
58  2013     1   1      544            545        -1     1004           1022   
59  2013     1   1      554            600        -6      812            837   

54 arr_delay carrier flight tailnum origin dest air_time distance hour minute  \
55        11      UA   1545  N14228    EWR  IAH      227     1400    5     15   
56        20      UA   1714  N24211    LGA  IAH      227     1416    5     29   
57        33      AA   1141  N619AA    JFK  MIA      160     1089    5     40   
58       -18      B6    725  N804JB    JFK  BQN      183     1576    5     45   
59       -25      DL    461  N668DN    LGA  ATL      116      762    6      0   

54         time_hour  
55  01-01-2013 05:00  
56  01-01-2013 05:00  
57  01-01-2013 05:00  
58  01-01-2013 05:00  
59  01-01-2013 06:00

In [86]:
planes.head()

tailnum    year                     type      manufacturer      model  \
0  N10156  2004.0  Fixed wing multi engine           EMBRAER  EMB-145XR   
1  N102UW  1998.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
2  N103US  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
3  N104UW  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
4  N10575  2002.0  Fixed wing multi engine           EMBRAER  EMB-145LR   

   engines  seats  speed     engine  
0        2     55    NaN  Turbo-fan  
1        2    182    NaN  Turbo-fan  
2        2    182    NaN  Turbo-fan  
3        2    182    NaN  Turbo-fan  
4        2     55    NaN  Turbo-fan

In [87]:
planes.columns

Index(['tailnum', 'year', 'type', 'manufacturer', 'model', 'engines', 'seats',
       'speed', 'engine'],
      dtype='object')

In [88]:
flights.columns

Index(['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay',
       'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight',
       'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute',
       'time_hour'],
      dtype='object', name=54)

In [89]:
result_15= flights.merge(planes[['tailnum', 'year', 'speed', 'seats']], how='left', on='tailnum')

In [90]:
result_15.columns

Index(['year_x', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay',
       'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight',
       'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute',
       'time_hour', 'year_y', 'speed', 'seats'],
      dtype='object')

**Here year_x belongs to flights and yea_y belongs to planes , we need to rename column names year_x with year,year_y with plane_year ,speed with plane_speed,seats with plane_seats**

In [91]:
result_15.rename(columns = {'year_x':'year','year_y':'plane_year','speed':'plane_speed','seats':'plane_seats'}, inplace = True)

In [92]:
result_15

year month day dep_time sched_dep_time dep_delay arr_time  \
0       2013     1   1      517            515         2      830   
1       2013     1   1      533            529         4      850   
2       2013     1   1      542            540         2      923   
3       2013     1   1      544            545        -1     1004   
4       2013     1   1      554            600        -6      812   
...      ...   ...  ..      ...            ...       ...      ...   
336771  2013     9  30      NaN           1455       NaN      NaN   
336772  2013     9  30      NaN           2200       NaN      NaN   
336773  2013     9  30      NaN           1210       NaN      NaN   
336774  2013     9  30      NaN           1159       NaN      NaN   
336775  2013     9  30      NaN            840       NaN      NaN   

       sched_arr_time arr_delay carrier  ... origin dest air_time distance  \
0                 819        11      UA  ...    EWR  IAH      227     1400   
1                 830        20      UA  ...    LGA  IAH      227     1416   
2                 850        33      AA  ...    JFK  MIA      160     1089   
3                1022       -18      B6  ...    JFK  BQN      183     1576   
4                 837       -25      DL  ...    LGA  ATL      116      762   
...               ...       ...     ...  ...    ...  ...      ...      ...   
336771           1634       NaN      9E  ...    JFK  DCA      NaN      213   
336772           2312       NaN      9E  ...    LGA  SYR      NaN      198   
336773           1330       NaN      MQ  ...    LGA  BNA      NaN      764   
336774           1344       NaN      MQ  ...    LGA  CLE      NaN      419   
336775           1020       NaN      MQ  ...    LGA  RDU      NaN      431   

       hour minute         time_hour plane_year plane_speed  plane_seats  
0         5     15  01-01-2013 05:00     1999.0         NaN        149.0  
1         5     29  01-01-2013 05:00     1998.0         NaN        149.0  
2         5     40  01-01-2013 05:00     1990.0         NaN        178.0  
3         5     45  01-01-2013 05:00     2012.0         NaN        200.0  
4         6      0  01-01-2013 06:00     1991.0         NaN        178.0  
...     ...    ...               ...        ...         ...          ...  
336771   14     55  30-09-2013 14:00        NaN         NaN          NaN  
336772   22      0  30-09-2013 22:00        NaN         NaN          NaN  
336773   12     10  30-09-2013 12:00        NaN         NaN          NaN  
336774   11     59  30-09-2013 11:00        NaN         NaN          NaN  
336775    8     40  30-09-2013 08:00        NaN         NaN          NaN  

[336776 rows x 22 columns]

In [93]:
result_15.columns

Index(['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay',
       'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight',
       'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute',
       'time_hour', 'plane_year', 'plane_speed', 'plane_seats'],
      dtype='object')

In [94]:
result_15.shape

(336776, 22)

#  16. SELECT planes.* , airlines.*  FROM (SELECT DISTINCT carrier,tailnumFROM flights) AS cartail INNER JOIN planes ON cartail.tailnum=planes.tailnum INNER JOIN airlines ON cartail.carrier=airlines.carrier

**two joins need to be done 
two primary keys are 1.tailnum and 2.carrier**

**select distinct pairs 'carrier' and 'tailnum' columns and store them in cartail**

In [95]:
flights.shape

(336776, 19)

In [96]:
cartail = flights[['carrier', 'tailnum']].drop_duplicates()

**compute inner join between cartail and planes**

In [97]:
inner_join_cartail_planes= pd.merge(cartail, planes, on='tailnum', how='inner')

**now compute innerjoin between above result and airlines**

In [98]:
result_16= pd.merge(inner_join_cartail_planes, airlines, on='carrier', how='inner')
result_16

carrier tailnum    year                     type    manufacturer  \
0         UA  N14228  1999.0  Fixed wing multi engine          BOEING   
1         UA  N24211  1998.0  Fixed wing multi engine          BOEING   
2         UA  N39463  2012.0  Fixed wing multi engine          BOEING   
3         UA  N29129  1998.0  Fixed wing multi engine          BOEING   
4         UA  N53441     NaN  Fixed wing multi engine          BOEING   
...      ...     ...     ...                      ...             ...   
3334      OO  N760SK  2005.0  Fixed wing multi engine  BOMBARDIER INC   
3335      OO  N766SK  2005.0  Fixed wing multi engine  BOMBARDIER INC   
3336      OO  N772SK  2005.0  Fixed wing multi engine  BOMBARDIER INC   
3337      OO  N776SK  2006.0  Fixed wing multi engine  BOMBARDIER INC   
3338      OO  N785SK  2009.0  Fixed wing multi engine  BOMBARDIER INC   

            model  engines  seats  speed     engine                   name  
0         737-824        2    149    NaN  Turbo-fan  United Air Lines Inc.  
1         737-824        2    149    NaN  Turbo-fan  United Air Lines Inc.  
2       737-924ER        2    191    NaN  Turbo-fan  United Air Lines Inc.  
3         757-224        2    178    NaN  Turbo-jet  United Air Lines Inc.  
4       737-924ER        2    191    NaN  Turbo-fan  United Air Lines Inc.  
...           ...      ...    ...    ...        ...                    ...  
3334  CL-600-2C10        2     80    NaN  Turbo-fan  SkyWest Airlines Inc.  
3335  CL-600-2C10        2     80    NaN  Turbo-fan  SkyWest Airlines Inc.  
3336  CL-600-2C10        2     80    NaN  Turbo-fan  SkyWest Airlines Inc.  
3337  CL-600-2C10        2     80    NaN  Turbo-fan  SkyWest Airlines Inc.  
3338  CL-600-2C10        2     80    NaN  Turbo-fan  SkyWest Airlines Inc.  

[3339 rows x 11 columns]

In [99]:
result_16.shape

(3339, 11)

**17.SELECT
flights2.
*, 
atemp, 
ahumid
FROM (
SELECT * FROM flights WHERE origin='EWR'
) AS flights2
LEFT
JOIN (
SELECT
year, month, day, 
AVG(temp) AS atemp, 
AVG(humid) AS ahumid
FROM weather
WHERE origin='EWR'
GROUP BY year, month, day**

# way 1

In [100]:
flights.shape

(336776, 19)

In [101]:
flights2 = flights[flights['origin'] == 'EWR']
flights2

54      year month day dep_time sched_dep_time dep_delay arr_time  \
55      2013     1   1      517            515         2      830   
60      2013     1   1      554            558        -4      740   
61      2013     1   1      555            600        -5      913   
68      2013     1   1      558            600        -2      923   
71      2013     1   1      559            600        -1      854   
...      ...   ...  ..      ...            ...       ...      ...   
336807  2013     9  30   2142.0           2129      13.0   2250.0   
336810  2013     9  30   2149.0           2156      -7.0   2245.0   
336811  2013     9  30   2150.0           2159      -9.0   2250.0   
336815  2013     9  30   2211.0           2059      72.0   2339.0   
336817  2013     9  30   2233.0           2113      80.0    112.0   

54     sched_arr_time arr_delay carrier flight tailnum origin dest air_time  \
55                819        11      UA   1545  N14228    EWR  IAH      227   
60                728        12      UA   1696  N39463    EWR  ORD      150   
61                854        19      B6    507  N516JB    EWR  FLL      158   
68                937       -14      UA   1124  N53441    EWR  SFO      361   
71                902        -8      UA   1187  N76515    EWR  LAS      337   
...               ...       ...     ...    ...     ...    ...  ...      ...   
336807           2239      11.0      EV   4509  N12957    EWR  PWM     47.0   
336810           2308     -23.0      UA    523  N813UA    EWR  BOS     37.0   
336811           2306     -16.0      EV   3842  N10575    EWR  MHT     39.0   
336815           2242      57.0      EV   4672  N12145    EWR  STL    120.0   
336817             30      42.0      UA    471  N578UA    EWR  SFO    318.0   

54     distance hour minute         time_hour  
55         1400    5     15  01-01-2013 05:00  
60          719    5     58  01-01-2013 05:00  
61         1065    6      0  01-01-2013 06:00  
68         2565    6      0  01-01-2013 06:00  
71         2227    6      0  01-01-2013 06:00  
...         ...  ...    ...               ...  
336807      284   21     29  30-09-2013 21:00  
336810      200   21     56  30-09-2013 21:00  
336811      209   21     59  30-09-2013 21:00  
336815      872   20     59  30-09-2013 20:00  
336817     2565   21     13  30-09-2013 21:00  

[120835 rows x 19 columns]

In [102]:
flights2 = flights2.merge(weather[weather['origin'] == 'EWR'].groupby(['year', 'month', 'day']).agg({'temp': 'mean', 'humid': 'mean'}).reset_index(), how='left')

In [103]:
flights2.columns

Index(['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay',
       'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight',
       'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute',
       'time_hour', 'temp', 'humid'],
      dtype='object')

In [104]:
flights2.rename(columns = {'temp':'atemp','humid':'ahumid'}, inplace = True)

In [105]:
flights2.columns

Index(['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay',
       'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight',
       'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute',
       'time_hour', 'atemp', 'ahumid'],
      dtype='object')

In [106]:
flights

54      year month day dep_time sched_dep_time dep_delay arr_time  \
55      2013     1   1      517            515         2      830   
56      2013     1   1      533            529         4      850   
57      2013     1   1      542            540         2      923   
58      2013     1   1      544            545        -1     1004   
59      2013     1   1      554            600        -6      812   
...      ...   ...  ..      ...            ...       ...      ...   
336826  2013     9  30      NaN           1455       NaN      NaN   
336827  2013     9  30      NaN           2200       NaN      NaN   
336828  2013     9  30      NaN           1210       NaN      NaN   
336829  2013     9  30      NaN           1159       NaN      NaN   
336830  2013     9  30      NaN            840       NaN      NaN   

54     sched_arr_time arr_delay carrier flight tailnum origin dest air_time  \
55                819        11      UA   1545  N14228    EWR  IAH      227   
56                830        20      UA   1714  N24211    LGA  IAH      227   
57                850        33      AA   1141  N619AA    JFK  MIA      160   
58               1022       -18      B6    725  N804JB    JFK  BQN      183   
59                837       -25      DL    461  N668DN    LGA  ATL      116   
...               ...       ...     ...    ...     ...    ...  ...      ...   
336826           1634       NaN      9E   3393     NaN    JFK  DCA      NaN   
336827           2312       NaN      9E   3525     NaN    LGA  SYR      NaN   
336828           1330       NaN      MQ   3461  N535MQ    LGA  BNA      NaN   
336829           1344       NaN      MQ   3572  N511MQ    LGA  CLE      NaN   
336830           1020       NaN      MQ   3531  N839MQ    LGA  RDU      NaN   

54     distance hour minute         time_hour  
55         1400    5     15  01-01-2013 05:00  
56         1416    5     29  01-01-2013 05:00  
57         1089    5     40  01-01-2013 05:00  
58         1576    5     45  01-01-2013 05:00  
59          762    6      0  01-01-2013 06:00  
...         ...  ...    ...               ...  
336826      213   14     55  30-09-2013 14:00  
336827      198   22      0  30-09-2013 22:00  
336828      764   12     10  30-09-2013 12:00  
336829      419   11     59  30-09-2013 11:00  
336830      431    8     40  30-09-2013 08:00  

[336776 rows x 19 columns]

In [107]:
flights2.shape

(120835, 21)

# way 2

In [114]:
second_query=weather[weather['origin'] == 'EWR'].groupby(['year', 'month', 'day']).agg({'temp': 'mean', 'humid': 'mean'}).reset_index()

In [115]:
second_query.rename(columns = {'temp':'atemp','humid':'ahumid'}, inplace = True)

In [116]:
second_query

year  month  day    atemp     ahumid
0    2013      1    1  38.4800  58.386087
1    2013      1    2  28.8350  47.786250
2    2013      1    3  29.4575  54.395833
3    2013      1    4  33.4775  55.880417
4    2013      1    5  36.7325  49.284583
..    ...    ...  ...      ...        ...
359  2013     12   26  31.0475  60.904167
360  2013     12   27  34.2425  56.687500
361  2013     12   28  39.1550  54.897500
362  2013     12   29  43.0475  67.602083
363  2013     12   30  38.9000  73.838750

[364 rows x 5 columns]

In [117]:
flights2 = flights[flights['origin'] == 'EWR']
flights2

54      year month day dep_time sched_dep_time dep_delay arr_time  \
55      2013     1   1      517            515         2      830   
60      2013     1   1      554            558        -4      740   
61      2013     1   1      555            600        -5      913   
68      2013     1   1      558            600        -2      923   
71      2013     1   1      559            600        -1      854   
...      ...   ...  ..      ...            ...       ...      ...   
336807  2013     9  30   2142.0           2129      13.0   2250.0   
336810  2013     9  30   2149.0           2156      -7.0   2245.0   
336811  2013     9  30   2150.0           2159      -9.0   2250.0   
336815  2013     9  30   2211.0           2059      72.0   2339.0   
336817  2013     9  30   2233.0           2113      80.0    112.0   

54     sched_arr_time arr_delay carrier flight tailnum origin dest air_time  \
55                819        11      UA   1545  N14228    EWR  IAH      227   
60                728        12      UA   1696  N39463    EWR  ORD      150   
61                854        19      B6    507  N516JB    EWR  FLL      158   
68                937       -14      UA   1124  N53441    EWR  SFO      361   
71                902        -8      UA   1187  N76515    EWR  LAS      337   
...               ...       ...     ...    ...     ...    ...  ...      ...   
336807           2239      11.0      EV   4509  N12957    EWR  PWM     47.0   
336810           2308     -23.0      UA    523  N813UA    EWR  BOS     37.0   
336811           2306     -16.0      EV   3842  N10575    EWR  MHT     39.0   
336815           2242      57.0      EV   4672  N12145    EWR  STL    120.0   
336817             30      42.0      UA    471  N578UA    EWR  SFO    318.0   

54     distance hour minute         time_hour  
55         1400    5     15  01-01-2013 05:00  
60          719    5     58  01-01-2013 05:00  
61         1065    6      0  01-01-2013 06:00  
68         2565    6      0  01-01-2013 06:00  
71         2227    6      0  01-01-2013 06:00  
...         ...  ...    ...               ...  
336807      284   21     29  30-09-2013 21:00  
336810      200   21     56  30-09-2013 21:00  
336811      209   21     59  30-09-2013 21:00  
336815      872   20     59  30-09-2013 20:00  
336817     2565   21     13  30-09-2013 21:00  

[120835 rows x 19 columns]

In [120]:
left_join= pd.merge(flights2,second_query,on=['year','month','day'], how='left')

In [121]:
left_join

year month day dep_time sched_dep_time dep_delay arr_time  \
0       2013     1   1      517            515         2      830   
1       2013     1   1      554            558        -4      740   
2       2013     1   1      555            600        -5      913   
3       2013     1   1      558            600        -2      923   
4       2013     1   1      559            600        -1      854   
...      ...   ...  ..      ...            ...       ...      ...   
120830  2013     9  30   2142.0           2129      13.0   2250.0   
120831  2013     9  30   2149.0           2156      -7.0   2245.0   
120832  2013     9  30   2150.0           2159      -9.0   2250.0   
120833  2013     9  30   2211.0           2059      72.0   2339.0   
120834  2013     9  30   2233.0           2113      80.0    112.0   

       sched_arr_time arr_delay carrier  ... tailnum origin dest air_time  \
0                 819        11      UA  ...  N14228    EWR  IAH      227   
1                 728        12      UA  ...  N39463    EWR  ORD      150   
2                 854        19      B6  ...  N516JB    EWR  FLL      158   
3                 937       -14      UA  ...  N53441    EWR  SFO      361   
4                 902        -8      UA  ...  N76515    EWR  LAS      337   
...               ...       ...     ...  ...     ...    ...  ...      ...   
120830           2239      11.0      EV  ...  N12957    EWR  PWM     47.0   
120831           2308     -23.0      UA  ...  N813UA    EWR  BOS     37.0   
120832           2306     -16.0      EV  ...  N10575    EWR  MHT     39.0   
120833           2242      57.0      EV  ...  N12145    EWR  STL    120.0   
120834             30      42.0      UA  ...  N578UA    EWR  SFO    318.0   

       distance hour minute         time_hour    atemp    ahumid  
0          1400    5     15  01-01-2013 05:00      NaN       NaN  
1           719    5     58  01-01-2013 05:00      NaN       NaN  
2          1065    6      0  01-01-2013 06:00      NaN       NaN  
3          2565    6      0  01-01-2013 06:00      NaN       NaN  
4          2227    6      0  01-01-2013 06:00      NaN       NaN  
...         ...  ...    ...               ...      ...       ...  
120830      284   21     29  30-09-2013 21:00  62.9075  69.80625  
120831      200   21     56  30-09-2013 21:00  62.9075  69.80625  
120832      209   21     59  30-09-2013 21:00  62.9075  69.80625  
120833      872   20     59  30-09-2013 20:00  62.9075  69.80625  
120834     2565   21     13  30-09-2013 21:00  62.9075  69.80625  

[120835 rows x 21 columns]